### DISCUSSION TITLE
Why the return type of "get()" command is JQuery<HTMLElement> if it can return intercepted requests?

### DISCUSSION THREAD

**Discussion by kryshenp** [cypress-io/cypress#30694] [Questions and Help]
We have a solution for verifying that no request(s) to certain endpoint(s) are made within a specific timeout after performing a certain action.
The solution works as expected and fulfills our requirements flawlessly.

However, we noticed that the return type of `cy.get()` is defined as `JQuery<HTMLElement>`.
Given the provided example, it seems that the return type of the get command should not always be `JQuery<HTMLElement>`, but also support intercepted requests and arrays of intercepted requests.

**Does anyone know why, in the Cypress API, the return type of the get command is always set to `JQuery<HTMLElement>`, even though use cases like the one mentioned below are possible?
Is this aspect documented or discussed anywhere?**

In `cypress/commands.ts` file. Notice that in `cy.get(${intercept.alias}.all).then(calls => {` the we work with `calls`  that are not the HTML element but intercepted requests. 
```
Cypress.Commands.add("failIfEndpointIsCalled", (endpointAlias, timeout, triggerFunction) => {
  const aliases = Array.isArray(endpointAlias) ? endpointAlias : [endpointAlias];
  const intercepts = aliases.map(alias => {
    return {alias, interceptedTimesBefore: 0};
  });

  intercepts.forEach(intercept => {
    cy.get(`${intercept.alias}.all`).then(calls => {
      intercepts;
      intercept.interceptedTimesBefore = calls.length;
    });
  });

  triggerFunction();
  cy.wait(timeout).then(() => {
    intercepts.forEach(intercept => {
      cy.get(`${intercept.alias}.all`, {timeout: 0})
        .should("have.length", intercept.interceptedTimesBefore, `Number of new requests to '${intercept.alias}' should be 0. `
          + `Total number of intercepted requests is expected to be ${intercept.interceptedTimesBefore}`);
    });
  });
})
```
Usage example in spec file. The test would fail if in 10 seconds after the click on the button there will be requests sent to endpoint1 or endpoint2.  `failIfEndpointIsCalled` command will accept a single intercepted request alias as an argument or an array of intercepted aliases. 
```
cy.intercept("**/endpoint1").as("endpoint1");
cy.intercept("**/endpoint2").as("endpoint2");
cy.failIfEndpointIsCalled(["@endpoint1", "@endpoint2"], 10000, () => {
      cy.get("button").click();
});
```

in cypress/types/index.d.ts file
```
declare namespace Cypress {
  interface Chainable {
    failIfEndpointIsCalled(
      endpointAlias: string | string[], 
      timeout: number, 
      triggerFunction: () => void
    ): Chainable<void>;
  }
}
```

<img width="604" alt="Screenshot 2024-12-02 at 17 26 07" src="https://github.com/user-attachments/assets/c08588ff-9975-438a-a7f2-5661327f5f53">

**Comment by portyu9** [comment]
Hi @kryshenp 

You were correct that the old TypeScript definition was misleading for aliases.

At runtime, `cy.get()` has always been able to retrieve more than DOM elements through aliases—for example:

```js
cy.get('@users')
cy.get('@myStub')
cy.get('@request')
cy.get('@request.all')
```

The `JQuery<HTMLElement>` result you were seeing was a **TypeScript typing limitation**, not the actual runtime behavior.

This has since been fixed.

Starting in **Cypress 15.17.0**, `cy.get()` has a separate alias overload for `@`-prefixed strings, so aliases no longer automatically resolve to:

```ts
Chainable<JQuery<HTMLElement>>
```

Current Cypress essentially has an overload like:

```ts
get<S = any>(
  alias: `@${string}`,
  options?: ...
): Chainable<S>
```

You can therefore provide the actual alias type explicitly when useful:

```ts
cy.get<Cypress.Interception[]>('@myRequest.all')
  .then((calls) => {
    expect(calls).to.have.length(2)
  })
```

One detail matters with dynamically constructed strings. If TypeScript only knows that your variable is a generic:

```ts
string
```

it may not be able to prove that the value starts with `@`.

Typing the alias accordingly helps:

```ts
const alias: `@${string}` = '@myRequest'

cy.get<Cypress.Interception[]>(`${alias}.all`)
```

So the short answer is: **yes, the old `JQuery<HTMLElement>` typing was incomplete, and Cypress fixed this in 15.17.0.**

Please mark this answer as accepted if it helped, thank you.
